Makefile

```diff
@@ -134,8 +134,24 @@ binary-containerized: vendor
 ###############################################################################
 
 ## Builds the code and runs all tests.
+# ensure we confirmation
+imagetag:
+
 ci: clean image check-copyright ut fv
 
+cd:
+ifndef CONFIRM
+	$(error CONFIRM is undefined - run using make <target> CONFIRM=true)
+endif
+ifndef BRANCH_NAME
+	$(error BRANCH_NAME is undefined - run using make <target> BRANCH_NAME=var or set an environment variable)
+endif
+	if [ -z "${PULL_REQUEST_NUMBER}" ]; then \
```

Comment:
I don't think we can rely on this PULL_REQUEST_NUMBER logic in the Makefile, unfortunately.

Semaphore doesn't give us an "ok-to-test" type command, which means anyone could submit a PR which drops this check and then they can push whatever images they want to the calico repositories without review. 

So, Semaphore needs to have the logic which says whether or not it should run the `push` target. It could be as simple as `if [ -z ${PULL_REQUEST_NUMBER} ]; then make cd; fi`, which is still a big improvement in terms of simplicity.

Reply:
> anyone could submit a PR which drops this check and then they can push whatever images they want to the calico repositories without review

You mean that someone could change the target to remove the test `if [ -z ...` and then when Semaphore runs `make ci` or `make cd` for the PR pre-merge CI it would push? Yes, I can see that.

The problem is that just about every current repo has some form of `make ci` or `make test` or `make build` or `make <something>` as part of the pre-merge PR CI run in Semaphore. All you would need to do is look in Semaphore, find any target, and change it.

This is a generalized problem. Other CI systems - travis, circle, Jenkins, etc. - all rely entirely on one file/dir that exists in the repo, and thus have the same issue.

I don't think moving that `if -z [ "{PULL_REQUEST_NUMBER}" ]` into Semaphore itself solves the security (or "dumb mistake") problem, since I could do the same thing in a `ci` target or any other target that Semaphore calls.

I think this is a general problem: how do you get non-master (i.e. pre-approved PR) CI to **not** do destructive things?

My view is that if you are trying to prevent that kind of thing, it is a security question. Make it so that nothing that someone puts into a PR can do anything destructive, i.e. eliminate credentials in Semaphore when running pre-merge PR CI.

Reply:
How _do_ people prevent destructive PRs in travis/circle/etc.?

Reply:
Aha. https://docs.travis-ci.com/user/pull-requests/#Pull-Requests-and-Security-Restrictions

```
A pull request sent from a fork of the upstream repository could be manipulated to expose environment variables. The upstream repository’s maintainer would have no protection against this attack, as pull requests can be sent by anyone who forks the repository on GitHub.

Travis CI makes encrypted variables and data available only to pull requests coming from the same repository. These are considered trustworthy, as only members with write access to the repository can send them.

Pull requests sent from forked repositories do not have access to encrypted variables or data.
```

Similar approach. Credentials - all of them - are unavailable when running PR CI, only post-merge. So you cannot do anything destructive (or expose them). This, I believe, is the correct approach: remove destructive credentials from CI runs that should not have them.

Reply:
Here is the link for Semaphore https://semaphoreci.com/docs/building-pull-requests.html#pull-requests-originating-from-forks

Nothing there about the creds for PR. I will email them.

Reply:
> The problem is that just about every current repo has some form of make ci or make test or make build or make <something> as part of the pre-merge PR CI run in Semaphore. All you would need to do is look in Semaphore, find any target, and change it.

blech, duh. Yes, not sure why I thought this case was special in some way. Clearly anyone can add an image push to anything.

I'll try to dig into semaphore a bit to see what they offer here.

Reply:
Unfortunately, yeah. I sent Semaphore a message, but maybe you will get a better response?

Reply:
The way we normally solve this is to have a setup command in semaphore that drops all the credentials so that a push would fail.  It's no good having semaphore check whether it should run the push target because someone could rewrite the build target to _try_ a push.

Reply:
I also heard back from Semaphore, same thing. At least they are consistent. :-)

Reply:
> setup command in semaphore that drops all the credentials so that a push would fail

At a certain point, you do need a push target. And even if you don't, as long as Semaphore is executing _any_ code, someone can modify that code in a branch, open a PR and boom. When it comes down to it, Semaphore just does this **wrong**, and the others do it right.